java-checks/src/main/java/org/sonar/java/checks/DoublePrefixOperatorCheck.java

```diff
@@ -38,35 +41,52 @@
     return ImmutableList.of(Tree.Kind.LOGICAL_COMPLEMENT, Tree.Kind.BITWISE_COMPLEMENT, Tree.Kind.UNARY_PLUS, Tree.Kind.UNARY_MINUS);
   }
 
-  static HashSet<ExpressionTree> prefixSet = new HashSet<>();
+  private Set<ExpressionTree> prefixSet = new HashSet<>();
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    prefixSet.clear();
+    super.scanFile(context);
+  }
 
   @Override
   public void visitNode(Tree tree) {
-    UnaryExpressionTree temp = (UnaryExpressionTree) tree;
-    if (usedForDoublePrefix(temp)) {
+    UnaryExpressionTree exprTree = (UnaryExpressionTree) tree;
+    if (alreadyReported(exprTree)) {
       return;
     }
-    ExpressionTree expr = ExpressionUtils.skipParentheses(temp.expression());
-    if (temp.is(expr.kind())) {
+    ExpressionTree expr = ExpressionUtils.skipParentheses(exprTree.expression());
+    if (caseHandlingDoubleTilde(expr)) {
+      return;
+    }
+    if (exprTree.is(expr.kind())) {
       prefixSet.add(expr);
-      reportIssue(temp.operatorToken(), ((UnaryExpressionTree) expr).operatorToken(), "Remove multiple operator prefixes.");
+      reportIssue(exprTree.operatorToken(), ((UnaryExpressionTree) expr).operatorToken(), "Remove multiple operator prefixes.");
     }
   }
 
-  private static boolean usedForDoublePrefix(UnaryExpressionTree tree) {
+  private boolean alreadyReported(UnaryExpressionTree tree) {
     if (prefixSet.contains(tree)) {
       return true;
     }
     Tree parent = tree;
-    while (parent != null) {
-      if (!parent.is(Tree.Kind.PARENTHESIZED_EXPRESSION, Tree.Kind.BITWISE_COMPLEMENT, Tree.Kind.LOGICAL_COMPLEMENT, Tree.Kind.UNARY_PLUS, Tree.Kind.UNARY_MINUS)) {
-        return false;
-      }
+    while (parent.is(Tree.Kind.PARENTHESIZED_EXPRESSION, Tree.Kind.BITWISE_COMPLEMENT, Tree.Kind.LOGICAL_COMPLEMENT, Tree.Kind.UNARY_PLUS, Tree.Kind.UNARY_MINUS)) {
       parent = parent.parent();
       if (prefixSet.contains(parent)) {
         return true;
       }
     }
     return false;
   }
+
+  private static boolean caseHandlingDoubleTilde(ExpressionTree expr) {
+    if (expr.is(Tree.Kind.BITWISE_COMPLEMENT)) {
+      ExpressionTree exprIdentifier = ExpressionUtils.skipParentheses(((UnaryExpressionTree) expr).expression());
+      if (exprIdentifier.is(Tree.Kind.IDENTIFIER) || exprIdentifier.getClass().equals(BinaryExpressionTreeImpl.class)) {
```

Comment:
Mmmh, if I agree with the first part, I don't with the second. 
`exprIdentifier.getClass().equals(BinaryExpressionTreeImpl.class)` is way too generic and does not cover the correct cases to me.

We want to spot the repetition of more than 2 bitwise complements in a row, and say nothing in case of only 2. So let's simplify this by rewriting the method:
(assuming you call from within the `if`)

```
private static boolean caseHandlingDoubleTilde(ExpressionTree expr) {
  if (expr.is(Tree.Kind.BITWISE_COMPLEMENT)) {
    if (ExpressionUtils.skipParentheses(((UnaryExpressionTree) expr).expression().is(Tree.Kind.BITWISE_COMPLEMENT) {
      // third '~' in a row, we can raise an issue
      return true;
    }
  }
  // only two, may be used for hash control
  return false;
}
```